cpp/src/strings/split/split.cu

```diff
@@ -271,544 +549,271 @@ struct whitespace_token_counter_fn
-      cudf::experimental::detail::round_up_pow2(info.token_size_sum, split_align);
-    auto const char_buf_ptr = memory_ptr;
-    memory_ptr += char_buf_size;
-    auto const offset_buf_ptr = reinterpret_cast<size_type*>(memory_ptr);
-
-    auto const d_str = d_strings.element<string_view>(info.idx);
-    size_type token_idx = 0;
-    size_type char_bytes_copied = 0;
-    size_type start_pos = 0;  // updates only if moving forward
-    auto end_pos = d_str.length();  // updates only if moving backward
-    while (token_idx < info.token_count - 1) {
-      auto const delimiter_pos =
-        dir == Dir::FORWARD ? d_str.find(d_delimiter, start_pos) :
-                              d_str.rfind(d_delimiter, start_pos, end_pos);
-      if (delimiter_pos != -1) {
-        auto const offset_size_pair =
-          compute_src_byte_offset_and_token_char_bytes<false>(
-            d_str, start_pos, end_pos, delimiter_pos);
-        if (dir == Dir::FORWARD) {
-          thrust::copy(
-            thrust::seq,
-            d_str.data() + offset_size_pair.first,
-            d_str.data() + offset_size_pair.first + offset_size_pair.second,
-            char_buf_ptr + char_bytes_copied);
-          offset_buf_ptr[token_idx] = char_bytes_copied;
-        }
-        else {
-          auto const char_buf_offset =
-            info.token_size_sum - char_bytes_copied - offset_size_pair.second;
-          thrust::copy(
-            thrust::seq,
-            d_str.data() + offset_size_pair.first,
-            d_str.data() + offset_size_pair.first + offset_size_pair.second,
-            char_buf_ptr + char_buf_offset);
-          offset_buf_ptr[info.token_count - 1 - token_idx] = char_buf_offset;
-        }
-        token_idx++;
-        char_bytes_copied += offset_size_pair.second;
-        if (dir == Dir::FORWARD) {
-          start_pos = delimiter_pos + d_delimiter.length();
-        }
-        else {
-          end_pos = delimiter_pos;
+struct whitespace_split_tokenizer_fn : base_whitespace_split_tokenizer
+{
+    /**
+     * @brief This will create tokens around each runs of whitespace characters.
+     *
+     * @param idx Index of the string to process
+     * @param column_count Number of output columns
+     * @param d_token_counts Token counts for each string
+     * @param d_all_tokens All output tokens for the strings column
+     */
+    __device__ void process_tokens(size_type idx,
+                                   size_type column_count,
+                                   size_type const* d_token_counts,
+                                   string_index_pair* d_all_tokens ) const
+    {
+        string_index_pair* d_tokens = d_all_tokens + (idx * column_count);
+        thrust::fill( thrust::seq, d_tokens, d_tokens + column_count, string_index_pair{nullptr,0} );
+        if( d_strings.is_null(idx) )
+            return;
+        string_view const d_str = d_strings.element<cudf::string_view>(idx);
+        if( d_str.empty() )
+            return;
```

Comment:
The `fill()` here serves two purposes. First is to fill all nulls for null and empty string case. Second is to fill the remainder of tokens with nulls for a string which has less delimiters than the output column count. Since this pulls in alot of inline code, I wanted to minimize the calls to it.
I profiled this code with and without the `fill()` here (for extreme case) and it had no significant impact on performance.